draft-frindell-moq-chat.md

```diff
@@ -55,115 +55,81 @@ receive messages published by others.
-charlie
-~~~
-
-Any Object in a Group with sequence number higher than 0 is a delta encoding,
-with new participants prefaced with a '+' character and participants that have
-left the chat prefaced with a '-' character.  Every delta encoding object MUST
-be applied for an endpoint to have a current list of participants.
-
-Delta Encoding Exmaple:
-
-~~~
-+daphne
--bob
-~~~
-
-Every username in the catalog MUST be unique.  If an endpoint receives a catalog
-or delta encoding that would result in the same username more than once, it MUST
-close the session with an error.
-
-Every object in a catalog group MUST be on a single stream, so that updates are
-delivered in order.
-
 # Chat Operation
 
-## MoQ Chat Server
+## MoQ Relay
 
-The protocol requires that one entity operate as a chat server.  The function
-of the server is to maintain the list of participants in the chat and publish
-catalog updates.
+The protocol requires a MoQ relay to act as a chat server.  The relay maintains
+the set of connected clients that have announced a chat track.
 
 ## Chat ID
 
 Every chat has a unique ID.  The ID is a string of arbitrary length and uniquely
-identifies the chat.  The creation of chat IDs and discovery of the server
-endpoint is out of the scope of this document.
+identifies the chat.  The creation of chat IDs and discovery of the relay is
+out of the scope of this document.
 
 ## Track Names
 
-The catalog for a given chat is available by subscribing to moq-chat/id.
+Each chat participant has a chat track.  The namespace of the track is
+("moq-chat", \<id\>, \<user-id\>, \<device-id\>, \<timestamp\>) and the name
+track is "chat".
 
-## Joining the Chat
+* id - the ID of the chatroom
+* user-id - the user ID
+* device-id - a unique identifier for each device for the user.  This allows
+              the same user to join the chat from multiple devices
+* timestamp - the timestamp in seconds when the track started.  This allows a
+              stateless client to start publishing without accidentally
+              overwriting a previously sent group and object.
 
-To join the chat a participant sends a SUBSCRIBE message to moq-chat/id in order
-to get the participant list and subsequent participant updates.  This SUBSCRIBE
-requires the AUTHORIZATION_INFO to contain the participant's ASCII string
-username.  It is RECOMMENDED the participant request the catalog starting from
-the beginning of the current Group. The moq-chat server will reply with an
-SUBSCRIBE_OK if joining was successful or SUBSCRIBE_ERROR if the join failed.
+## Joining the Chat
 
-The participant also sends an ANNOUNCE message to the server with a track
-namespace of moq-chat/id/participant/username, and an empty Track Name.
+To join the chat a participant sends a SUBSCRIBE_NAMESPACE message to the relay
+with a namesapce prefix ("moq-chat", \<id\>) in order to receive all ANNOUNCE
+message from each participant in the chat and UNANNOUNCE messages when
+participants leave.
```

Comment:
Not sure if this an MOQT question or a chat question, but does the relay forward SUBSCRIBE_NAMESPACE to everyone that has ANNOUNCEd something within the namespace, or does the relay keep track of ANNOUNCE so it can send ANNOUNCE in response to later SUBSCRIBE_NAMESPACE? What state does the relay keep?

MoQT refers to the "publisher" as the responder, which can include the relay. So I guess the relay is tracking all the ANNOUNCEs it's seen, and maybe has to prune them when the session goes away?

Reply:
draft-06 doesn't specify if SUBSCRIBE_NAMESPACE is forwarded through a relay.  The use case would be relay-to-relay.

A relay already needs to track a mapping from announcements to the sessions that announced them.  It uses this table to match incoming subscribes to the correct session(s).

To implement SUBSCRIBE_NAMESPACE in moxygen I did have to create a kind of tree-structure organized by namespace tuple, with a list of subscribed sessions at each level, and a session that has announced that namespace, if any.  I suppose when I support > 1 announcer per namespace I need that to be a list.

And when a session disappears you need to remove old announcements, and subscribed namespaces (and subscriptions).

Reply:
OK. More text in the doc is always helpful, even if it's non-normative.